### ISSUE TITLE
server: rule(MD051) frontmatter check works for CLI but not for server

### ISSUE THREAD

**Issue by eread** [rvben/rumdl#792]
### rumdl version

0.2.49

### Installation method

mise

### Bug description

`rumdl` rule MD051 has a `check-frontmatter` configuration option (https://github.com/rvben/rumdl/blob/6099a6c11d687480bdaf82f01d8f9695d5998570/docs/md051.md#check-frontmatter) for checking link fragments that are used in frontmatter.

This seems to work when `rumdl check` is run, but not `rumdl server`.

### Steps to reproduce

1. Create a file called `test.md` with the following contents:
   ```markdown
   ---
   title: Heading 1
   link: 'test.md#heading'
   ---
   
   This is a Markdown file.
   ```
1. Create a `.rumdl.toml` file with the following contents:
   ```markdown
   [global]
   enable = ["MD051"]
   
   [MD051]
   check-frontmatter = true
   ```
1. Run `rumdl check test.md`. `rumdl` returns:
   ```shell
   test.md:3:8: [MD051] Link fragment 'heading' not found in 'test.md'
   ```

### Expected behavior

Within an editor that invokes `rumdl server`, the problem on line three should be flagged.

### Actual behavior

Within an editor that invokes `rumdl server`, the problem on line three is not flagged.

### Configuration

```toml
[global]
enable = ["MD051"]

[MD051]
check-frontmatter = true
```

### Operating system

macOS

**Comment by rvben** [comment]
Fixed in [v0.2.51](https://github.com/rvben/rumdl/releases/tag/v0.2.51).

The cause was broader than the option: the language server hand-rolled its own workspace index instead of asking each rule what a file contributes, so everything a rule's configuration decided about indexing was invisible to it. `check-frontmatter` was the visible symptom, and your reproduction is now a test.

The same hand-rolled copy had four other defects that came out with it, all of which would have bitten you eventually:

- it generated every anchor GitHub-style whatever the flavor, and never deduplicated them;
- it omitted HTML and attribute anchors entirely, so `<a id="...">` and `{#custom-id}` targets read as missing;
- it never exported the inline-disable state that cross-file checks honor, so a `rumdl-disable` comment was honored on the command line and ignored in the editor;
- it anchored a cross-file diagnostic at the link destination where MD051 anchors it at the link, so the editor's squiggle sat on a different span than the CLI's column.

`build_file_index` now delegates to the same builder `rumdl check` uses, over the rules whose cross-file scope is the workspace, so the editor and the command line answer from one index rather than two.

Two more went in on top of it, both reachable from your setup: the server took document paths straight from the URI while canonicalizing workspace roots at initialize, so a file reached through a symlinked ancestor could never match its own index entry and every cross-file feature was silently dead for it; and frontmatter values were being offered as find-references results, which rename could not rewrite and therefore left stale.

Thank you for the report, and for the precise reproduction - "works under `check`, not under `server`" is exactly the framing that made it findable.

**Comment by eread** [comment]
@rvben Thanks for the fix! I can see it now in my editor.

One little caveat. I can't see it in my editor until I make a change. So after initially loading the file in the scenario above, nothing is flagged.

As soon as I make a change on the file in the scenario above, the MD051 error is shown.

**Comment by rvben** [comment]
Thanks for the feedback @eread. I'll have another look to see if that can be improved further.

**Comment by eread** [comment]
> Thanks for the feedback [@eread](https://github.com/eread). I'll have another look to see if that can be improved further.

thanks @rvben. I see you've made a relevant commit to `main`, and I'll confirm the fix when it ships 🙇🏻
